### PR TITLE
Consolidate templates for Fiori elements FPM

### PR DESCRIPTION
**Pull request by tobiasqueck**
The goal is to provide templates for adding custom extensions to a Fiori elements for OData v4 application that can be used in the SAP Fiori tools (in guided development as well as page modeler) that follow the guidelines provided in the Fiori elements flexible programming model explorer at https://ui5.sap.com/test-resources/sap/fe/core/fpmExplorer/index.html#/customElements/customElementsOverview

The current state is a draft implementation to get a better understanding what is needed for #73

### CODE REVIEW

**Comment by nlunets** [comment]
I would say 1.93

File: packages/templates/fev4-fpm/src/page/index.ts
Comment (815are) [inline packages/templates/fev4-fpm/src/page/index.ts:81]: I am not sure how actual issue would be, but in case if FCL is enabled, then target property should be an array, like in following FCL example - https://sapui5.hana.ondemand.com/#/topic/7f651316d396400d88835adb0d021532

Comment (tobiasqueck) [reply packages/templates/fev4-fpm/src/page/index.ts:81]: Very good point, the current implementation would not work with FCL. I guess there is not way around checking the routing configuration, and if it is FCL, update all routes.

Comment (nlunets) [reply packages/templates/fev4-fpm/src/page/index.ts:81]: I think technically the array notation also works for non FCL

Comment (tobiasqueck) [reply packages/templates/fev4-fpm/src/page/index.ts]: @815are could you check again, please?

Comment (815are) [reply packages/templates/fev4-fpm/src/page/index.ts]: @tobiasqueck 
I think that logic is not correct - https://github.com/SAP/open-ux-tools/blob/fev4-fpm/packages/templates/fev4-fpm/src/page/index.ts#L52

I think it should be something like - 
```
if (sourceRoute?.target.constructor === Array) {
            newRoute.target = [...sourceRoute.target, (newRoute as Ui5Route).name];       
        } else {
            newRoute.target = newRoute.name;
        }
```

Proof is that in snapshot we see identical `target` for both routes - https://github.com/SAP/open-ux-tools/blob/fev4-fpm/packages/templates/fev4-fpm/test/unit/__snapshots__/page.test.ts.snap#L20-L31

```
"routes":  [
        {
          "name": "TestObjectPage",
          "pattern": "object/{key}",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
        {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
      ]
```

As far as I know it should be:
```
"routes":  [
        {
          "name": "TestObjectPage",
          "pattern": "object/{key}",
          "target": [
            "TestObjectPage"
          ],
        },
        {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
      ]
```

Comment (815are) [reply packages/templates/fev4-fpm/src/page/index.ts]: Also in `spec` we have limitation for array length, something like:
```
// 1-3 level navigation is placed like [`1st Page`, `2nd Page`, `3rd Page`]
// 4+ level navigation is placed as fullscreen - [`4th+n Page`]
route.target = level <= 2 ? [...targetPages, route.target] : [route.target];
```
As I remember reason is that FCL supports only 3 columns - in result we do not need array with 4 columns.

**Comment by 815are** [comment]
> I would say 1.93

is it regarding `component` or `controller` approach? In `tool-suite` we have two templates using `controller` approach:
1. For versions above 1.88 with `sap/fe/core/PageController`
2. For versions below 1.88 with `sap/ui/core/mvc/Controller` instead of `sap/fe/core/PageController`

Main explanation was that `sap/fe/core/PageController` is available since `1.88` - https://sapui5.hana.ondemand.com/#/api/sap.fe.core.PageController

**Comment by stefanschreck** [comment]
> @zdravko-georgiev thanks a lot for your feedback.
> 
> I like your proposal about making the target folder configurable. I will enhance the API and use the current static setting as the default if no target folder is provided. Regarding the target name, changing the name (and therefore changing it also for new pages generated with App Modeller) shouldn't cause any compatibility problems, correct? The naming pattern used here is in sync with the naming pattern used in the writers for Fiori elements pages `<entity>PageType`, so it would create some consistency there.

@tobiasqueck 
Yes, the naming pattern should not be a problem, we just need the flexibility w.r.t the folder -  to reach compatibility with existing S/4 apps as well as what was generated by the tools-suite so far.

File: packages/fev4-fpm-writer/src/action/index.ts
Comment (stefanschreck) [inline packages/fev4-fpm-writer/src/action/index.ts:88]: here we also need more flexibility w.r.t the path

File: packages/fev4-fpm-writer/src/column/index.ts
Comment (stefanschreck) [inline packages/fev4-fpm-writer/src/column/index.ts:39]: more flexibility needed for the path

File: packages/fev4-fpm-writer/src/page/index.ts
Comment (stefanschreck) [inline packages/fev4-fpm-writer/src/page/index.ts:124]: same here. The ext folder should be the fallback if the caller does not supply a dedicated path.

**Review by stefanschreck — CHANGES_REQUESTED** [review]

File: packages/fev4-fpm-writer/src/page/index.ts
Comment (stefanschreck) [reply packages/fev4-fpm-writer/src/page/index.ts:124]: But I just saw https://github.com/SAP/open-ux-tools/pull/162 for the same

File: packages/fev4-fpm-writer/src/action/index.ts
Comment (stefanschreck) [reply packages/fev4-fpm-writer/src/action/index.ts:88]: But I just saw #162 for the same

File: packages/fev4-fpm-writer/templates/action/manifest.json
Comment (nlunets) [inline packages/fev4-fpm-writer/templates/action/manifest.json:7]: Technically this override now applies to all the ObjectPage, ideally we should strive to make this extension specific to one page

File: packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml
Comment (nlunets) [inline packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml:4]: Why do you wrap the Page in an App ? This is not needed nor is it natural

File: packages/fev4-fpm-writer/templates/page/1.84/ext/View.xml
Comment (nlunets) [inline packages/fev4-fpm-writer/templates/page/1.84/ext/View.xml:4]: Same remark for App here as on the other one

File: packages/fev4-fpm-writer/templates/page/1.84/ext/Controller.js
Comment (nlunets) [inline packages/fev4-fpm-writer/templates/page/1.84/ext/Controller.js:3]: supprot -> support

File: packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml
Comment (nlunets) [inline packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml]: Adding a vertical Layout as the basic sample we have for a column is really not nice, i'm not even sure they should be used.

Why don't you just add the control directly ?

File: packages/fev4-fpm-writer/templates/action/manifest.json
Comment (nlunets) [reply packages/fev4-fpm-writer/templates/action/manifest.json:7]: Such as this sap.fe.templates.ObjectPage.ObjectPageController#sap.fe.core.fpmExplorer::Default

File: packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml
Comment (tobiasqueck) [reply packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml]: Hmm ... I think this was taken from https://ui5.sap.com/test-resources/sap/fe/core/fpmExplorer/index.html#/customElements/customElementsOverview/customPageContent. If I knew the author of it ...

Comment (nlunets) [reply packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml]: image
Doesn't seem like it :)

File: packages/fev4-fpm-writer/templates/action/manifest.json
Comment (nlunets) [reply packages/fev4-fpm-writer/templates/action/manifest.json]: Docu link : https://sapui5.hana.ondemand.com/#/topic/358cf2598d71462b8ac2bd8c944efbfd

File: packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml
Comment (stefanschreck) [reply packages/fev4-fpm-writer/templates/page/1.94/ext/View.xml]: At least we use this pattern for the V2 custom pages all the time and it works, but we can try a simplification for V4

File: packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml
Comment (stefanschreck) [reply packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml]: That's what we do in V2, an that's what is documented for V4: https://ui5.sap.com/#/topic/d525522c1bf54672ae4e02d66b38e60c.html

Comment (nlunets) [reply packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml]: I'll double check with the colleagues and potentially ask for an update in docu for this

File: packages/fe-fpm-writer/src/action/types.ts
Comment (nlunets) [inline packages/fe-fpm-writer/src/action/types.ts:7]: items is a navProperty here not something specific

File: packages/fe-fpm-writer/src/action/types.ts
Comment (nlunets) [inline packages/fe-fpm-writer/src/action/types.ts:6]: FieldGroup refers more to a form than a facet, have a look at https://ui5.sap.com/#/topic/7619517a92414e27b71f02094bd08d06 for details on the different action extension that are supported

File: packages/fev4-fpm-writer/templates/action/manifest.json
Comment (tobiasqueck) [reply packages/fev4-fpm-writer/templates/action/manifest.json]: @nlunets like this: https://github.com/SAP/open-ux-tools/pull/39/commits/d9b50b4ddf1e008c24a5be1369bcca1ad638170a?

File: packages/fe-fpm-writer/src/action/types.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/action/types.ts]: @nlunets https://github.com/SAP/open-ux-tools/pull/39/commits/81370928578f68709d8457a9973d9d289c94a66e better?

Comment (nlunets) [reply packages/fe-fpm-writer/src/action/types.ts]: yes

File: packages/fe-fpm-writer/templates/page/1.94/manifest.json
Comment (zdravko-georgiev) [inline packages/fe-fpm-writer/templates/page/1.94/manifest.json:4]: This overrides the minUI5Version of my application. Can we remove it from the template?
image

Comment (tobiasqueck) [reply packages/fe-fpm-writer/templates/page/1.94/manifest.json:4]: Oha ... that is wrong. I wouldn't remove it but it should only override if the current `minUI5Version` is smaller.

Comment (zdravko-georgiev) [reply packages/fe-fpm-writer/templates/page/1.94/manifest.json:4]: Or we have a validation thats checks if the app version is correct and throws an exception when I call e.g. `generateCustomPage`?

File: packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml
Comment (tobiasqueck) [reply packages/fev4-fpm-writer/templates/column/ext/CustomColumnFragment.xml]: @nlunets as discussed,  the `VerticalLayout` has been removed.

File: packages/fe-fpm-writer/templates/page/1.94/manifest.json
Comment (tobiasqueck) [reply packages/fe-fpm-writer/templates/page/1.94/manifest.json]: Actually, we have the optional input parameter allowing the caller to define the compatibility requirement for UI5, so we should not set this value, it is the responsibility of the consumer. Removed with https://github.com/SAP/open-ux-tools/pull/39/commits/b4bac19ab15e43a892eafaafbfdc15fc5e8d229d

File: packages/fe-fpm-writer/templates/page/1.84/manifest.json
Comment (zdravko-georgiev) [inline packages/fe-fpm-writer/templates/page/1.84/manifest.json:5]: @tobiasqueck can we remove also this hardcoded version?

File: packages/fe-fpm-writer/src/action/index.ts
Comment (stefanschreck) [inline packages/fe-fpm-writer/src/action/index.ts:85]: There should be an option to run generateCustomAction without this step. For scenarios that want to reuse the same event handler multiple times:
a) if you want to have the same extension action at different places of the UI
b) if you wat to use the same event handler for multiple actions (like in the FPM explorer example)

File: packages/fe-fpm-writer/src/page/index.ts
Comment (stefanschreck) [inline packages/fe-fpm-writer/src/page/index.ts:47]: This validation makes sense for the other extension types as well

File: packages/fe-fpm-writer/src/page/index.ts
Comment (stefanschreck) [inline packages/fe-fpm-writer/src/page/index.ts:26]: sonarlint indicates that `as string` is superfluous

File: packages/fe-fpm-writer/src/page/index.ts
Comment (stefanschreck) [inline packages/fe-fpm-writer/src/page/index.ts:22]: if sourceRoute cannot be found, the pattern will comprise undefined. Shouldn`t we rather throw an exception?

File: packages/fe-fpm-writer/src/action/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/action/index.ts:85]: Agree, also the coding for event handling of a custom column and a custom action should be the same. Will take care of it.

File: packages/fe-fpm-writer/src/page/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/page/index.ts]: df10458

File: packages/fe-fpm-writer/src/page/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/page/index.ts]: 3cbaa5b

File: packages/fe-fpm-writer/src/page/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/page/index.ts:37]: 90a595f

Comment (stefanschreck) [reply packages/fe-fpm-writer/src/page/index.ts:37]: Fine

File: packages/fe-fpm-writer/src/action/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/action/index.ts]: c866f3d

File: packages/fe-fpm-writer/templates/page/1.84/manifest.json
Comment (tobiasqueck) [reply packages/fe-fpm-writer/templates/page/1.84/manifest.json]: 6e537f9

File: packages/fe-fpm-writer/src/action/index.ts
Comment (stefanschreck) [reply packages/fe-fpm-writer/src/action/index.ts]: Great, thanks

**Review by stefanschreck — DISMISSED** [review]
OK for me, but a merge from`main` seems to be required

**Review by stefanschreck — APPROVED** [review]

**Comment by tobiasqueck** [comment]
Before merging, I would need an approval for each of the following aspects

- [ ] Integration into SAP Fiori tools application modeller (@stefanschreck @zdravko-georgiev)
- [ ] correctness for generated elements (@nlunets )
- [ ] standards and requirements for code in `open-ux-tools` (@devinea @ullasholla )

**Review by zdravko-georgiev — DISMISSED** [review]

File: .vscode/launch.json
Comment (stefanschreck) [inline .vscode/launch.json:58]: path must be changed now

**Review by stefanschreck — CHANGES_REQUESTED** [review]

File: .vscode/launch.json
Comment (tobiasqueck) [reply .vscode/launch.json]: 30be824

**Review by nlunets — APPROVED** [review]
Looks good content wise

**Review by stefanschreck — DISMISSED** [review]

**Review by nlunets — APPROVED** [review]

**Review by devinea — COMMENTED** [review]
@tobiasqueck some comments included.

Thanks

File: packages/fe-fpm-writer/src/column/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/column/index.ts:23]: For the column case, we could do the versioning slightly different and remove the duplications, however, in discussions with @stefanschreck we agreed that we want to try to keep the version specific templates separate so that they don't accidentally break when changing the template for the latest UI5 version i.e. once committed these old-version-templates shall not be touched anymore.
But, to be honest, I am not to sure yet, whether this is the best approach for versioning. Can we do it like this for now and change the approach if we have a better idea?

File: packages/fe-fpm-writer/src/common/validate.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/common/validate.ts:37]: Can we please do this in a separate PR. This is a chore we need to do across all modules in this repo. We have barely localized error messages (https://github.com/SAP/open-ux-tools/search?q=new+Error)

File: packages/fe-fpm-writer/src/common/types.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/common/types.ts:77]: Agree, but this should also go into a different PR to consolidate things across modules. I'd prefer keeping this PR focus on the module in question.

File: packages/fe-fpm-writer/src/action/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/action/index.ts:79]: We do the same in the `odata-service-writer` :(  (https://github.com/SAP/open-ux-tools/search?q=webapp+join)
I will create a separate issue for this one (and also list the other issues you mentioned that require corrections across modules)

File: packages/fe-fpm-writer/src/column/index.ts
Comment (devinea) [reply packages/fe-fpm-writer/src/column/index.ts:23]: Sure 👍

File: packages/fe-fpm-writer/src/action/index.ts
Comment (devinea) [reply packages/fe-fpm-writer/src/action/index.ts:79]: Yeah sure we can handle it in separate issue. 
Reason i raised it here is that this will more likely be applied to an existing project.

File: packages/fe-fpm-writer/src/common/types.ts
Comment (devinea) [reply packages/fe-fpm-writer/src/common/types.ts:77]: 👍

File: packages/fe-fpm-writer/src/common/validate.ts
Comment (devinea) [reply packages/fe-fpm-writer/src/common/validate.ts:37]: OK.

File: packages/fe-fpm-writer/src/action/index.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/action/index.ts:79]: I fully agree, I didn't want to open that can of worms yet :(

File: packages/fe-fpm-writer/src/common/types.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/common/types.ts:77]: https://github.com/SAP/open-ux-tools/issues/170

File: packages/fe-fpm-writer/src/common/validate.ts
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/common/validate.ts:37]: https://github.com/SAP/open-ux-tools/issues/170

File: packages/fe-fpm-writer/package.json
Comment (tobiasqueck) [reply packages/fe-fpm-writer/package.json:24]: Consolidating the version should be done in a separate PR as part of #170. We should get all modules to `9.3.0` since we are using the `fs.dump` function in many tests which is working in `9.0.0` but not documented (and part of the types)

Comment (devinea) [reply packages/fe-fpm-writer/package.json:24]: As discussed this is fine. We can update the other to 9.3.0 in a week or two.

**Review by devinea — APPROVED** [review]
Thanks @tobiasqueck lgtm and follow-up TBI created https://github.com/SAP/open-ux-tools/issues/170

**Review by kranthie-sap — APPROVED** [review]

**Pull request merged**
